src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -128,12 +151,52 @@ absl::flat_hash_set<NodeID> GcsResourceScheduler::FilterCandidateNodes(
   return result;
 }
 
-const std::vector<ResourceRequest> &GcsResourceScheduler::SortRequiredResources(
+std::vector<int> GcsResourceScheduler::SortRequiredResources(
     const std::vector<ResourceRequest> &required_resources) {
-  // TODO(ffbin): A bundle may require special resources, such as GPU. We need to
-  // schedule bundles with special resource requirements first, which will be implemented
-  // in the next pr.
-  return required_resources;
+  std::vector<int> sorted_index(required_resources.size());
+  std::iota(sorted_index.begin(), sorted_index.end(), 0);
+  std::sort(sorted_index.begin(), sorted_index.end(), [&](int b_idx, int a_idx) {
+    auto a = required_resources[a_idx];
+    auto b = required_resources[b_idx];
+
+    RAY_CHECK(a.predefined_resources.size() == (int)PredefinedResources_MAX);
+    RAY_CHECK(b.predefined_resources.size() == (int)PredefinedResources_MAX);
+
+    // Make sure that resources are always sorted in the same order
+    std::set<uint64_t> extra_resources_set;
+    for (auto r : a.custom_resources) {
+      extra_resources_set.insert(r.first);
+    }
+    for (auto r : b.custom_resources) {
+      extra_resources_set.insert(r.first);
+    }
+
+    // TODO (jon-chuang): the exact resource priority defined here needs to be revisted.
+
+    // Notes: This is a comparator for sorting in c++. We return true if a < b based on a
+    // resource at the given level of priority. If tied, we attempt to resolve based on
+    // the resource at the next level of priority
```

Comment:
can you comment the high level approach?

GPU > custom > other predefined
+ bigger values first